[lldb] Add 'modify' type watchpoints, make it default

jasonmolenda opened the pull request:
Watchpoints in lldb can be either 'read', 'write', or 'read/write'.  This is exposing the actual behavior of hardware watchpoints.  gdb has a different behavior: a "write" type watchpoint only stops when the watched memory region *changes*.

A user is using a watchpoint for one of three reasons:

1. Want to find what is changing/corrupting this memory.
3. Want to find what is reading from this memory.
2. Want to find what is writing to this memory.

I believe (1) is the most common use case for watchpoints, and it currently can't be done in lldb -- the user needs to continue every time the same value is written to the watched-memory manually.  I think gdb's behavior is the correct one.  There are some use cases where a developer wants to find every function that writes/reads to/from a memory region, regardless of value, I want to still allow that functionality.

This is also a bit of groundwork for my large watchpoint support proposal
https://discourse.llvm.org/t/rfc-large-watchpoint-support-in-lldb/72116 where I will be adding support for AArch64 MASK watchpoints which watch power-of-2 memory regions.  A user might ask to watch 24 bytes, and a MASK watchpoint stub can do this with a 32-byte MASK watchpoint if it is properly aligned. And we need to ignore writes to the final 8 bytes of that watched region, and not show those hits to the user.

This patch adds a new 'modify' watchpoint type and it is the default.

The part of this patch I think is the most questionable is the
SBTarget::CreateWatchpoint(addr, size, bool read, bool write, error) which I currently change the meaning of write==True to be Modify.  This is part of making 'modify' the default watchpoint type - a driver is likely setting write==True for its watchpoints.

I chatted with Alex and Jim about this a little and I'm not sure how I should solve this for real.  Add a new SBTarget::CreateWatchpoint API with a third `bool modify` argument in addition to the old one? Add a new SBWatchpointOptions class in case we want to add more types of watchpoints?  I'm not sure what other types of watchpoints we'd want to hardcode in our support.  We have conditional expressions or python callbacks that can do more unique operations.

There's more work here - the SB API needs to give the driver a way to specify all three possible types of watchpoint ('write' and 'modify' being exclusive of the other) - but I'd love to hear what other people think we should do for this API especially.

jasonmolenda committed:
[lldb] Add 'modify' type watchpoints, make it default

Watchpoints in lldb can be either 'read', 'write', or
'read/write'.  This is exposing the actual behavior of
hardware watchpoints.  gdb has a different behavior:
a "write" type watchpoint only stops when the watched
memory region *changes*.

A user is using a watchpoint for one of three reasons:

1. Want to find what is changing/corrupting this memory.
3. Want to find what is reading from this memory.
2. Want to find what is writing to this memory.

I believe (1) is the most common use case for watchpoints,
and it currently can't be done in lldb -- the user needs
to continue every time the same value is written to the
watched-memory manually.  I think gdb's behavior is the
correct one.  There are some use cases where a developer
wants to find every function that writes/reads to/from a memory
region, regardless of value, I want to still allow that
functionality.

This is also a bit of groundwork for my large watchpoint
support proposal
https://discourse.llvm.org/t/rfc-large-watchpoint-support-in-lldb/72116
where I will be adding support for AArch64 MASK watchpoints
which watch power-of-2 memory regions.  A user might ask
to watch 24 bytes, and a MASK watchpoint stub can do this
with a 32-byte MASK watchpoint if it is properly aligned.
And we need to ignore writes to the final 8 bytes of that
watched region, and not show those hits to the user.

This patch adds a new 'modify' watchpoint type and it
is the default.

The part of this patch I think is the most questionable
is the
SBTarget::CreateWatchpoint(addr, size, bool read, bool write, error)
which I currently change the meaning of write==True to be
Modify.  This is part of making 'modify' the default watchpoint
type - a driver is likely setting write==True for its watchpoints.

I chatted with Alex and Jim about this a little and I'm not sure
how I should solve this for real.  Add a new SBTarget::CreateWatchpoint
API with a third `bool modify` argument in addition to the old one?
Add a new SBWatchpointOptions class in case we want to add more
types of watchpoints?  I'm not sure what other types of watchpoints
we'd want to hardcode in our support.  We have conditional expressions
or python callbacks that can do more unique operations.

There's more work here - the SB API needs to give the driver a way
to specify all three possible types of watchpoint ('write' and
'modify' being exclusive of the other) - but I'd love to hear what
other people think we should do for this API especially.

jasonmolenda commented:
Just to be clear, I need to find a solution for SBTarget::CreateWatchpoint before this PR is ready to land.  I'm inclined towards adding a third bool parameter 'modify', but I'm not sure that's the best choice.

JDevlieghere commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -211,6 +212,35 @@ bool Watchpoint::CaptureWatchedValue(const ExecutionContext &exe_ctx) {
+  if (!m_type.IsValid())
+    return true;
+
+  ConstString watch_name("$__lldb__watch_value");
```
Seems like you can pass this directly to `ValueObjectMemory::Create` as a string literal and make @bulbazord happy?

bulbazord replied:
I've been reading through ValueObject and I'm not sure yet what the implications of removing ConstString from ValueObject are. If and when we can remove it from ValueObject, we'll be able to clean up this one too, so I would say don't worry about it right now.

As a suggestion, if you anticipate `WatchedValueReportable` being called many times, it would be less work to create that ConstString once as a `static` rather than creating one on each invocation.

JDevlieghere commented on lldb/include/lldb/Interpreter/OptionGroupWatchpoint.h:
```diff
@@ -31,13 +31,15 @@ class OptionGroupWatchpoint : public OptionGroup {
   void OptionParsingStarting(ExecutionContext *execution_context) override;
 
   // Note:
-  // eWatchRead == LLDB_WATCH_TYPE_READ; and
+  // eWatchRead == LLDB_WATCH_TYPE_READ
   // eWatchWrite == LLDB_WATCH_TYPE_WRITE
+  // eWatchModify == LLDB_WATCH_TYPE_MODIFY
```
These should be Doxygen (`///`) comments.

JDevlieghere commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -211,6 +212,35 @@ bool Watchpoint::CaptureWatchedValue(const ExecutionContext &exe_ctx) {
+      exe_ctx.GetBestExecutionContextScope(), watch_name.GetStringRef(),
+      watch_address, m_type);
+  newest_valueobj_sp = newest_valueobj_sp->CreateConstantValue(watch_name);
+  DataExtractor new_data;
+  DataExtractor old_data;
+  Status error;
+  newest_valueobj_sp->GetData(new_data, error);
+  m_new_value_sp->GetData(old_data, error);
```
Maybe this is just personal preference but I like to define the local as close to its use as possible. This feels a bit "C99".

JDevlieghere commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -211,6 +212,35 @@ bool Watchpoint::CaptureWatchedValue(const ExecutionContext &exe_ctx) {
+  newest_valueobj_sp = newest_valueobj_sp->CreateConstantValue(watch_name);
+  DataExtractor new_data;
+  DataExtractor old_data;
+  Status error;
```
Should we do anything with the error here? At least log it if something went wrong?

bulbazord replied:
+1, don't want to drop an error on the floor if we have one.

JDevlieghere commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -211,6 +212,35 @@ bool Watchpoint::CaptureWatchedValue(const ExecutionContext &exe_ctx) {
+  if (memcmp(new_data.GetDataStart(), old_data.GetDataStart(),
+             old_data.GetByteSize()) == 0)
+    return false; // Value has not changed, user requested modify watchpoint
+  else
+    return true;
```
else-after-return

JDevlieghere commented on lldb/source/Target/StopInfo.cpp:
```diff
@@ -982,6 +982,13 @@ class StopInfoWatchpoint : public StopInfo {
+
+        // Don't stop if the watched region value is unmodified, and
+        // this is a Modify-type watchpoint.
+        if (m_should_stop && !wp_sp->WatchedValueReportable(exe_ctx)) {
+          m_should_stop = false;
+        }
```
LLVM [says](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) don’t use braces on simple single-statement bodies.

clayborg commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -211,6 +212,35 @@ bool Watchpoint::CaptureWatchedValue(const ExecutionContext &exe_ctx) {
+  Status error;
+  newest_valueobj_sp->GetData(new_data, error);
+  m_new_value_sp->GetData(old_data, error);
+
```
Do we need to update m_new_value_sp with the value in newst_valueobj_sp somewhere so we detect when things change again?

jasonmolenda replied:
In StopInfoWatchpoint::PerformAction() we decide if we're `m_should_stop` and if so, we call `Watchpoint::CaptureWatchedValue` (one of the things we do before that call is compare the value and decide if we're going to stop or not).  CaptureWatchedValue copies the "new" value into the m_old_value_sp and then creates a new constant value object for m_new_value_sp.  I copied that same way of creating the const ValueObject in this method; it's the method before this one in Watchpoint.cpp.  So the copying of the new value to the previous value is already handled by that code in CaptureWatchedValue.

jasonmolenda replied:
Should add, because I'm sure someone will wonder about this:  Yes, I'm creating a constant ValueObject with the watched memory here, and then throwing that ValueObject away.  And if we decide to stop, we'll create the constant ValueObject again with the same memory, to show it to the user.  I'm relying on the memory read cache to only read that memory from the inferior process once at a single stop, otherwise this double-creation of a ValueObject would be a real perf hit to watchpoints.

clayborg commented on lldb/test/API/functionalities/watchpoint/modify-watchpoints/main.c:
```diff
@@ -0,0 +1,17 @@
+#include <stdint.h>
+int main() {
+  int value = 5;
```
If you take the address of "value" then it will always be in memory. Here we might be just getting lucky that the compiler choses to put this on the stack at -O0.

clayborg commented on lldb/test/API/functionalities/watchpoint/modify-watchpoints/main.c:
```diff
@@ -0,0 +1,17 @@
+  value = 5;
+  value = 5;
+  value = 10;
+  value = 10;
```
Can we check that it stops when the value changes from 10 to something else to ensure we detect all watchpoint changes and we can ensure that m_new_value_sp and newest_valueobj_sp are doing their jobs correctly? We need to make sure we stop on all changes, here we are testing on change

clayborg replied:
We could change it back to 5 to ensure that the watchpoint detects when we change from 5 to 10 (and stops) and then back to 5 (it should stop again because it changed from 10 to 5).

clayborg reviewed (commented):
Since users probably want a behavior like GDB by default, one idea is to just add a setting:
```
(llldb) settings show target.watchpoint-stop-on-write [always|changed]
```
When watchpoints are set, they just read the current value for this. Most of your internal code in this patch can stay the same. This allows LLDB to do what most people want by default and would require no changes to the command line command options or any API changes.

clayborg commented:
> Just to be clear, I need to find a solution for SBTarget::CreateWatchpoint before this PR is ready to land. I'm inclined towards adding a third bool parameter 'modify', but I'm not sure that's the best choice.

Anytime we keep adding more options to an API call we have two options:
- keep overloading new variants but leaving the old functions there for API backward compat
  - `SBWatchpoint SBTarget::WatchAddress(lldb::addr_t addr, size_t size, bool read, bool write, SBError &error);`
  - `SBWatchpoint SBTarget::WatchAddress(lldb::addr_t addr, size_t size, bool read, bool write, bool modify, SBError &error);`
- Add a new SBWatchpointOptions class that can you add as many accessors as you want to in the future, and then add a new SBTarget::CreateWatchpoint(SBWatchpointOptions options). This allows you to add any new options to the class without needing a new SBTarget::CreateWatchpoint variant in the future.
  - `lldb::SBWatchpoint WatchAddress(SBWatchpointOptions options, SBError &error);` 

If you think you will add more watchpoint options at some point in the future, go the SBWatchpointOptions route.

jasonmolenda commented:
> > Just to be clear, I need to find a solution for SBTarget::CreateWatchpoint before this PR is ready to land. I'm inclined towards adding a third bool parameter 'modify', but I'm not sure that's the best choice.
> 
> Anytime we keep adding more options to an API call we have two options:

Yeah exactly, I wasn't restating the options as clearly in that last comment, that's exactly what I meant.  I can't think of other options we would want to build in to a Watchpoint at creation time - but the argument for SBWatchpointOptions is that ten-years-in-the-future-Jason may want another flag.

Probably the best thing is to look at the SBTarget BreakpointCreate* methods; there are a dozen different methods for the different types of breakpoints you might want to create (address breakpoint, file & line breakpoint, symbol name breakpoint etc). An interesting aside is that none of the SB API methods take a flag for whether breakpoint should be set using a software or hardware breakpoint.  Jonas added that feature to debugserver a few years ago for x86_64 and aarch64, and I think he added the `target.require-hardware-breakpoint` setting then.  Otherwise the only way to set a hardware breakpoint is through the commandline `breakpoint set` command.

For a watchpoint, we only have SBTarget::WatchAddress, which takes and address and size.  If we were trying to follow the breakpoint API naming style, we would add SBTarget::WatchpointCreateByAddress, SBTarget::WatchpointCreateByVariable, and SBTarget::WatchpointCreateByExpression methods.  All of them would take the same read/write/modify flags, which might be the strongest argument for an options class even if it seems a little bit much for a few bools.

jasonmolenda commented:
Maybe I should just use a flag param, and adopt the BreakpointCreate style naming convention, even if there's only this one API right now (we should provide all three so driver authors don't need to duplicate the work, especially for a variable path).  e.g.

```
   SBWatchpoint SBTarget::WatchpointCreateByAddress(addr_t address, size_t size, uint32_t access_flags, SBError &error);
```

with `eWatchpointAccess{Read,Write,Modify}` flags defined.

bulbazord commented on lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:
```diff
@@ -3109,14 +3109,15 @@ static GDBStoppointType GetGDBStoppointType(Watchpoint *wp) {
-  assert(watch_read || watch_write);
-  if (watch_read && watch_write)
+  // watch_read, watch_write, watch_modify cannot all be false.
+  assert(watch_read || watch_write || watch_modify);
```
It might be worth putting that comment into the assertion so it's a little easier to understand at first glance?

```
  assert((watch_read || watch_write || watch_modify) && "watch_read, watch_write, watch_modify cannot all be false.");
  ```

clayborg commented:
> > > Just to be clear, I need to find a solution for SBTarget::CreateWatchpoint before this PR is ready to land. I'm inclined towards adding a third bool parameter 'modify', but I'm not sure that's the best choice.
> > 
> > 
> > Anytime we keep adding more options to an API call we have two options:
> 
> Yeah exactly, I wasn't restating the options as clearly in that last comment, that's exactly what I meant. I can't think of other options we would want to build in to a Watchpoint at creation time - but the argument for SBWatchpointOptions is that ten-years-in-the-future-Jason may want another flag.

The SBTarget::Launch(SBLaunchInfo) and SBTarget::Attach(SBAttachInfo) are some of the best example of using an options class. Breakpoints never tried to use an options class, we just kept adding new APIs which makes our API a bit messier, but we must keep APIs after they are added. So I would vote for the SBWatchpointOptions method. 

In the future we might have move complex ways to track watchpoints or want to track multiple address ranges with a single watchpoint? 
> 
> Probably the best thing is to look at the SBTarget BreakpointCreate* methods; there are a dozen different methods for the different types of breakpoints you might want to create (address breakpoint, file & line breakpoint, symbol name breakpoint etc). An interesting aside is that none of the SB API methods take a flag for whether breakpoint should be set using a software or hardware breakpoint. Jonas added that feature to debugserver a few years ago for x86_64 and aarch64, and I think he added the `target.require-hardware-breakpoint` setting then. Otherwise the only way to set a hardware breakpoint is through the commandline `breakpoint set` command.

Yeah, we probably didn't want to add yet another API call to just add access to the flag.
> 
> For a watchpoint, we only have `SBTarget::WatchAddress`, which takes and address and size. If we were trying to follow the breakpoint API naming style, we would add `SBTarget::WatchpointCreateByAddress`, `SBTarget::WatchpointCreateByVariable`, and `SBTarget::WatchpointCreateByExpression` methods. All of them would take the same read/write/modify flags, which might be the strongest argument for an options class even if it seems a little bit much for a few bools.

Yeah, it isn't too hard and it allows us great flexibility in the future. I have a patch coming for improving the saving of core files that allows users to specify a custom memory region list. I am creating a new options class named SBProcessSaveCoreOptions for the SBProcess::SaveCore(SBProcessSaveCoreOptions options) function as we can always add more flags to this class.

jasonmolenda commented:
> ```
>    SBWatchpoint SBTarget::WatchpointCreateByAddress(addr_t address, size_t size, uint32_t access_flags, SBError &error);
> ```
> 
> with `eWatchpointAccess{Read,Write,Modify}` flags defined.

@bulbazord what do you think about this suggestion?  Would you still prefer an Options class?

clayborg commented:
> > ```
> >    SBWatchpoint SBTarget::WatchpointCreateByAddress(addr_t address, size_t size, uint32_t access_flags, SBError &error);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with `eWatchpointAccess{Read,Write,Modify}` flags defined.
> 
> @bulbazord what do you think about this suggestion? Would you still prefer an Options class?

If we are going to add an overload I would suggest keeping with just adding a "bool modify" as it is more clear an usable. The options does seem like overkill for just one bool I admit, it just depends on what kind of options we might want to watchpoints in the future. If this is the last change to watchpoints, then add a new API. If we think we will add more options at some point (try to think if the new fancy watchpoints Jason is about to add support for might need more options?) then do the Options class route.

jasonmolenda commented:
> > > ```
> > >    SBWatchpoint SBTarget::WatchpointCreateByAddress(addr_t address, size_t size, uint32_t access_flags, SBError &error);
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > with `eWatchpointAccess{Read,Write,Modify}` flags defined.
> > 
> > 
> > @bulbazord what do you think about this suggestion? Would you still prefer an Options class?
> 
> If we are going to add an overload I would suggest keeping with just adding a "bool modify" as it is more clear an usable. The options does seem like overkill for just one bool I admit, it just depends on what kind of options we might want to watchpoints in the future. If this is the last change to watchpoints, then add a new API. If we think we will add more options at some point (try to think if the new fancy watchpoints Jason is about to add support for might need more options?) then do the Options class route.

You and Alex both preferred adding an Options class to pass in to this (and future WatchpointCreate API) so I'll write that up for my next revision of this PR, I didn't see your earlier message talking about your preference for that when I ping'ed Alex on their opinion.

You were talking about how you think the proliferation of BreakpointCreate API was not ideal.  Do you think there should be a `SBTarget::WatchpointCreate(SBWatchpointSpec, SBError &error)` and you would specify (1) addr + size, (2) variable, (3) expression + size, as well as the access flags, for the Watchpoint?  I don't mind the separate SBTarget methods approach for Breakpoints, it seems like six of one, half a dozen of the other to me.

It's probably best to just go with `SBTarget::WatchpointCreateByAddress` and add a new SBWatchpointOptions with the access flags only.

bulbazord commented:
> > ```
> >    SBWatchpoint SBTarget::WatchpointCreateByAddress(addr_t address, size_t size, uint32_t access_flags, SBError &error);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with `eWatchpointAccess{Read,Write,Modify}` flags defined.
> 
> @bulbazord what do you think about this suggestion? Would you still prefer an Options class?

I'm alright with an Options class. What I would like to see is flexibility instead of adding new overloads every time we may want to add new levers and knobs to change Watchpoint behavior.

jimingham commented:
I'm not sure I see the benefit of the SBWatchpointSpec including the way the watched region is specified.  That's just moving the proliferation of API's from SBTarget::WatchpointCreate*() to the SBWatchpointSpec constructor. We'd have to have:

SBWatchpointSpec::SBWatchpointSpec(SBAddress addr, int flags);
SBWatchpointSpec::SBWatchpointSpec(SBValue value, int flags);
SBWatchpointSpec::SBWatchpointSpec(const char *expr, int flags);

And then users would do:

wp_spec = lldb.SBWatchpointSpec("foo + 5", lldb.eWatchpointTypeModify)
wp = target.WatchpointCreate(wp_spec)

So that's really one more API plus some boilerplate to construct the options.  If an SBWatchpointSpec were the sort of thing you'd reuse, there might be some sense in making that a separate entity, but I don't see that for watchpoints.  Otherwise, I don't think it adds much.

Jim


> On Sep 14, 2023, at 3:12 PM, Jason Molenda ***@***.***> wrote:
> 
> 
>    SBWatchpoint SBTarget::WatchpointCreateByAddress(addr_t address, size_t size, uint32_t access_flags, SBError &error);
> with eWatchpointAccess{Read,Write,Modify} flags defined.
> 
> @bulbazord <https://github.com/bulbazord> what do you think about this suggestion? Would you still prefer an Options class?
> 
> If we are going to add an overload I would suggest keeping with just adding a "bool modify" as it is more clear an usable. The options does seem like overkill for just one bool I admit, it just depends on what kind of options we might want to watchpoints in the future. If this is the last change to watchpoints, then add a new API. If we think we will add more options at some point (try to think if the new fancy watchpoints Jason is about to add support for might need more options?) then do the Options class route.
> 
> You and Alex both preferred adding an Options class to pass in to this (and future WatchpointCreate API) so I'll write that up for my next revision of this PR, I didn't see your earlier message talking about your preference for that when I ping'ed Alex on their opinion.
> 
> You were talking about how you think the proliferation of BreakpointCreate API was not ideal. Do you think there should be a SBTarget::WatchpointCreate(SBWatchpointSpec, SBError &error) and you would specify (1) addr + size, (2) variable, (3) expression + size, as well as the access flags, for the Watchpoint? I don't mind the separate SBTarget methods approach for Breakpoints, it seems like six of one, half a dozen of the other to me.
> 
> It's probably best to just go with SBTarget::WatchpointCreateByAddress and add a new SBWatchpointOptions with the access flags only.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/66308#issuecomment-1720222526>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW4XRSYKJBWO3X4RAY3X2N6NVANCNFSM6AAAAAA4XKAUOQ>.
> You are receiving this because your review was requested.
>

clayborg commented:
> You and Alex both preferred adding an Options class to pass in to this (and future WatchpointCreate API) so I'll write that up for my next revision of this PR, I didn't see your earlier message talking about your preference for that when I ping'ed Alex on their opinion.
> 
> You were talking about how you think the proliferation of BreakpointCreate API was not ideal. Do you think there should be a `SBTarget::WatchpointCreate(SBWatchpointSpec, SBError &error)` and you would specify (1) addr + size, (2) variable, (3) expression + size, as well as the access flags, for the Watchpoint? I don't mind the separate SBTarget methods approach for Breakpoints, it seems like six of one, half a dozen of the other to me.
> 
> It's probably best to just go with `SBTarget::WatchpointCreateByAddress` and add a new SBWatchpointOptions with the access flags only.

I do like the SBWatchpointSpec idea as then we don't need any overloads for the different kinds of watchpoints (address, value, etc). One thing I really like about the options/spec approach is if you are writing a new command in lldb that sets a watchpoint, and if that command has options in the python, then it is easy to populate the options/spec gradually and then set a watchpoint at the end. So the flow can be:
- Create a default "SBWatchPointSpec spec;"
- Parse the options as they come in and call accessors on the "spec" object to setup the watchpoint
- Call SBTarget::WatchpointCreate(spec)

That being said I understand that it makes using python to quickly set watchpoints a bit more of a pain, so I am not voting 100% in any direction, just throwing stuff out there. I already have to check the APIs everytime I use python since we have many overloads on functions already, not a big deal if we add more. One thing to think about though is to make sure there aren't and default python arguments that would make it hard to call the existing API or the new API. There is some build warning when I build right now about a method being covered up due to default args when swig parses the header files + .i files that might be having this issue. So if we go the route of overloading we just need to make sure we avoid that issue in python.

The other way we can configure this is to add an accessor to the SBWatchpoint object itself for non common options like "stop on all writes even when not modified". So we could leave the create API alone, and then add a method like:
```
bool SBWatchpoint::SetStopOnAllWrites(bool b);
```
So we do the right thing by only stopping on a write the modifes by default and then we can further configure the breakpoint after it has been created.

Looking at the 6 overloads for setting breakpoints by file and line, 7 for breakpoint by name(s), 6 for regex, I still vote either the options route or adding an accessor on the SBWatchpoint object after it is created in lieu of adding overloads.

jasonmolenda committed:
Update changes for first round of feedback

Add a SBWatchpointOptions class for specifying the read/write/modify
type of a watchpoint.

Mark SBTarget::WatchAddress as deprecated, add
SBTarget::WatchpointCreateByAddress.

When a user specifies a "read-write" watchpoint, don't override the
"write" half of that to modify.  In this case, the behavior the
user probably wants is to see every access to this memory range,
not just modifications of the value.

Address first round of feedback from Jonas, Greg, Alex.

clayborg commented on lldb/include/lldb/API/SBWatchpointOptions.h:
```diff
@@ -0,0 +1,44 @@
+
+  void SetWatchpointTypeRead(bool read);
+  bool GetWatchpointTypeRead() const;
+
```
Add header doc here to explain that write mode will stop on any write even if the value doesn't change

clayborg commented on lldb/include/lldb/API/SBWatchpointOptions.h:
```diff
@@ -0,0 +1,44 @@
+  ~SBWatchpointOptions();
+
+  const SBWatchpointOptions &operator=(const lldb::SBWatchpointOptions &rhs);
+
```
add a header doc explaining this will stop any time memory is read from. Might want to mention if this works along with "modify" at all to only stop when reading a value and that value has changed?

jasonmolenda replied:
I don't think we can do read+modify, actually. I fixed a subtle bug in this latest commit in Watchpoint::WatchedValueReportable() where we know that a watched memory region may have been accessed.  If it's read+modify, we don't know if it's been read from, or written to (value mutating or not).  Unless we emulate the instruction that was executed, it could be a read, or a write with the same value, or it could be an access *near* the watched region which triggered the stop.  (this latter happening with large watchpoint support where watching a 32 byte region may actually need to watch 32 bytes, or on an AArch64 processor in streaming SVE mode when a write is *near* the watched region)

clayborg commented on lldb/include/lldb/API/SBWatchpointOptions.h:
```diff
@@ -0,0 +1,44 @@
+  void SetWatchpointTypeWrite(bool write);
+  bool GetWatchpointTypeWrite() const;
+
+  void SetWatchpointTypeModify(bool modify);
```
add a header doc explaining this does and if it can be used along with "read" or "write" or only "write"?

clayborg commented on lldb/source/API/SBTarget.cpp:
```diff
@@ -1326,23 +1326,35 @@ lldb::SBWatchpoint SBTarget::WatchAddress(lldb::addr_t addr, size_t size,
 
+  SBWatchpointOptions options;
+  options.SetWatchpointTypeRead(read);
+  options.SetWatchpointTypeModify(write);
```
do we want to rename this argument from "write" to "modify" to clarify what will happen? Variable names are not included in the mangled name, so we can rename as long as we don't change the type and not affect the API backwards compat. If we do rename, then update the header file as well

jasonmolenda replied:
Good point.

clayborg commented on lldb/source/API/SBWatchpointOptions.cpp:
```diff
@@ -0,0 +1,67 @@
+using namespace lldb;
+using namespace lldb_private;
+
+class WatchpointOptionsImpl {
+public:
+  bool m_read = false;
+  bool m_write = false;
+  bool m_modify = false;
+};
```
This could be moved to Watchpoint.h and then used in our internal API as well if you like that idea.

jasonmolenda replied:
Yeah I originally wrote that little class in Watchpoint.h and then I moved it into the SBWatchpointOptions because I wasn't using it anywhere else.  tbh I thought the class for the three bools was overkill right now but I agree that there's long-term value in making the SB API more extensible.  But for the lldb_private API, I'm less worried about using three bools for now.  If it does become more than that, an Options class would be a thing to adopt.

clayborg replied:
Yeah, we can do anything internally, so it doesn't really matter what we do inside as we can always change it at any point as long as the public API is solid.

clayborg commented on lldb/test/API/functionalities/watchpoint/unaligned-spanning-two-dwords/TestUnalignedSpanningDwords.py:
```diff
@@ -46,7 +46,9 @@ def test_unaligned_watchpoint(self):
+        wp_opts = lldb.SBWatchpointOptions()
+        wp_opts.SetWatchpointTypeModify(True)
+        wp = target.WatchpointCreateByAddress(a_bytebuf_6_addr, 4, wp_opts, err)
         self.assertTrue(err.Success())
```
No need to change this test if it still uses the old API. Revert?

jasonmolenda replied:
I was marking `SBTarget::WatchAddress` as deprecated, so I thought updating the API tests to use the current API was the best choice.  It might be good to keep one using the old API though, to catch a break in it.

clayborg commented on lldb/test/API/python_api/default-constructor/sb_target.py:
```diff
@@ -53,7 +53,10 @@ def fuzz_obj(obj):
     obj.GetTriple()
     error = lldb.SBError()
-    obj.WatchAddress(123, 8, True, True, error)
+    wp_opts = lldb.SBWatchpointOptions()
+    wp_opts.SetWatchpointTypeRead(True)
+    wp_opts.SetWatchpointTypeModify(True)
+    obj.WatchpointCreateByAddress(123, 8, wp_opts, error)
```
Revert and keep old API usage?

clayborg commented on lldb/include/lldb/API/SBWatchpointOptions.h:
```diff
@@ -0,0 +1,44 @@
+
+  const SBWatchpointOptions &operator=(const lldb::SBWatchpointOptions &rhs);
+
+  void SetWatchpointTypeRead(bool read);
+  bool GetWatchpointTypeRead() const;
+
+  void SetWatchpointTypeWrite(bool write);
+  bool GetWatchpointTypeWrite() const;
+
+  void SetWatchpointTypeModify(bool modify);
+  bool GetWatchpointTypeModify() const;
```
Since we are changing the API, would it be more clear to just have an enum and accessor to make it clear when we still stop will happen during writes? We can keep the SetWatchpointTypeRead(bool), but for the write maybe something like:
```
// Put this into lldb-enumerations.h
namespace lldb { 
enum class WatchpointWriteType {
  Disabled, // Don't stop when the data being watched is written to
  Always, // Stop on any write access to the the value being watched even if value doesn't change
  OnModify // Stop only on write access that modifies the value being watched
};
} // namespace lldb
```
The we have just one accessor for the write control like:
```
lldb::WatchpointWriteType GetWatchpointWriteType();
void SetWathpointWriteType(lldb::WatchpointWriteType watch_type);
```
The current API in this class might be confusing to the user as to what needs to be set:
- If I want to stop when the value is modfied, do I just set "modify" or also need to set "write"?
- Can I set "read" and "modify"? (we said no on this, but it might still be confusing to users as they might think they can)
- what does setting "read", "write" and "modify" do?

jasonmolenda replied:
This is a good idea.  I was a unhappy with "read", "write", and "modify" as if they were all independently selectable values or something, but hadn't thought of a better way of expressing write/modify as a type of write watchpoint.

clayborg reviewed (commented):
Overall looks pretty solid. Just need to nail down the headerdoc for the "modify" accesors and clarify that it only works in combination with "write" and probably change the old API from "bool read, bool write" to be "bool read, bool modify" to clearly indicate what will happen. Or checkout the suggestion for the lldb::WatchpointWriteType enum and let me know what you think

jasonmolenda committed:
Address Greg's feedback, most important, enum for type of write

Add a new `WatchpointWriteType` to lldb-enumerations,
change SBWatchpointOptions::SetWatchpointTypeWrite() to
take that enum as its argument, eWatchpointWriteTypeOnModify,
eWatchpointWriteTypeAlways, or eWatchpointWriteTypeDisabled.  This
clarifies that a watchpoint is either a "read" watchpoint, a "write"
watchpoint, or both.  And if it's a "write" watchpoint, it is either
'always' or 'modify'.

clayborg commented on lldb/bindings/interface/SBWatchpointOptionsDocstrings.i:
```diff
@@ -6,11 +6,7 @@
 %feature("docstring", "Gets whether the watchpoint should stop on read accesses."
 ) lldb::SBWatchpointOptions::GetWatchpointTypeRead;
-%feature("docstring", "Sets whether the watchpoint should stop on write accesses."
+%feature("docstring", "Sets whether the watchpoint should stop on write accesses. eWatchpointWriteTypeOnModify is the most commonly useful mode, where lldb will stop when the watched value has changed. eWatchpointWriteTypeAlways will stop on any write to the watched region, and on some targets there can false watchpoint stops where memory near the watched region was written, and lldb cannot detec ...
```
Do we need to state "and on some targets there can false watchpoint stops where memory near the watched region was written, and lldb cannot detect that it is a spurious stop."? Can we remove that? Seems like information that most users wouldn't need to know?

jasonmolenda replied:
Yeah I'll remove that text.  It's going to be part of my large watchpoints enhancement -- on stubs that support it on AArch64, we'll use a masking watchpoint, so you can watch a 48 byte object but we'll do it by watching 64 bytes of memory.  for a straight-up "write" watchpoint, we can't tell whether the 48 bytes were written to with the same value, or the 16 bytes we aren't intending to watch were written to.  Also with AArch64 in SVE Streaming Mode, the address matching is at lower accuracy, so a write near a watchpoint can match the watchpoint, even if it didn't actually access the memory range.

jasonmolenda replied:
(and the "modify" watchpoint type solves both of these, because we can only report watchpoint hits where the watched memory region changes)

clayborg reviewed (approved):
Thanks for the changes! Looks great

jimingham commented:
That watchpoints can be inaccurate is probably news to a lot of debugger users, and knowing that might save somebody a fruitless hour wondering "how that variable could be being modified here..."  But I don't think most folks are likely to see the API doc string.  Might be more useful to mention this in the "watch set" help.

Jikm

> On Sep 18, 2023, at 6:38 PM, Greg Clayton ***@***.***> wrote:
> 
> 
> @clayborg approved this pull request.
> 
> Thanks for the changes! Looks great
> 
> In lldb/bindings/interface/SBWatchpointOptionsDocstrings.i <https://github.com/llvm/llvm-project/pull/66308#discussion_r1329453193>:
> 
> > @@ -6,11 +6,7 @@
>  ) lldb::SBWatchpointOptions::SetWatchpointTypeRead;
>  %feature("docstring", "Gets whether the watchpoint should stop on read accesses."
>  ) lldb::SBWatchpointOptions::GetWatchpointTypeRead;
> -%feature("docstring", "Sets whether the watchpoint should stop on write accesses."
> +%feature("docstring", "Sets whether the watchpoint should stop on write accesses. eWatchpointWriteTypeOnModify is the most commonly useful mode, where lldb will stop when the watched value has changed. eWatchpointWriteTypeAlways will stop on any write to the watched region, and on some targets there can false watchpoint stops where memory near the watched region was written, and lldb cannot detect that it is a spurious stop."
> Do we need to state "and on some targets there can false watchpoint stops where memory near the watched region was written, and lldb cannot detect that it is a spurious stop."? Can we remove that? Seems like information that most users wouldn't need to know?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/66308#pullrequestreview-1632250909>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVWYQDHG3WAIHK77KF2DX3DZRNANCNFSM6AAAAAA4XKAUOQ>.
> You are receiving this because your review was requested.
>

jasonmolenda committed:
Don't detail the corner case of false watchpoint stops

in the SBWatchpointOptions docstring.

DavidSpickett commented:
I've reverted this due to failures on Arm/AArch64 Linux, well, one test failing, one passing unexpectedly. Likely they are expecting the watchpoint to go off even when not modified. If I can fix them I'll do that and reland this.

DavidSpickett commented:
I've updated the tests and relanded but there is still some stuff to check. The tagged address test is working fine but the default watchpoint type is different between the 2 commands, which seems wrong at first glance.

For the other one see https://github.com/llvm/llvm-project/issues/26405 for some context. I guess somewhere we need to check for a read/write or modify break on the location before allowing the user to set another one, since that's where the test used to fail when it tried to place the second watchpoint. Since the globals are all chars and their addresses when aligned down were the same, I think. See if you can figure that one out.

DavidSpickett commented:
Where the "commands" are:
```
watchpoint set expression -s 4 -- tagged_ptr
watchpoint set variable global_var
```
They default to different types.

jasonmolenda commented:
Thanks @DavidSpickett for looking into these.  Let me looked into the tagged ptr watchpoint issue right now, that sounds like a bug.  One of the changes I made in the second landing of my patch was to the TestStepOverWatchpoint C file, which had three 1-byte globals, one which it would put a `read` watchpoint on, and second that it would put a `write` watchpoint on.  lldb will send these as separate watchpoint requests (Z2,addr,1).  If the variables were allocated in consecutive bytes, which seems likely, and we use a doubleword resolution watchpoint, then the remote stub would need to realize that we're watching for 'read' accesses on one byte, and 'write' accesses on a different byte within that doubleword, and use a single hardware watchpoint register with read+write access enabled.

I don't know if lldb-server would handle this correctly, but debugserver wouldn't.  And it's not what the test case intends to be testing.  When I re-landed my patch, I changed the variables to `long` so they're more likely to be the size of the hardware watchpoint granule and aligned to that alignment.  I wonder why it was still failing, i'll have to look more.

jasonmolenda commented:
FWIW I've found that `watch set expression -s <n>` will ignore the size specified if there is a clang type of the expression, it seems. :(

jasonmolenda commented:
Ah, I see the bug where `watch set expression` ends up marked as `rw`.  CommandObjectWatchpointSetExpression::DoExecute passes `OptionGroupWatchpoint::eWatchModify` as the type to Target::CreateWatchpoint (which takes that as a uint32_t).  It passes that `uint32_t kind` to Watchpoint::SetWatchpointType which is expecting the `uint32_t kind` to be a bitfield with `LLDB_WATCH_TYPE_{READ,WRITE,MODIFY}`.  `OptionGroupWatchpoint::eWatchModify` had value 3, so we get bits 1 and 2 set here, which is `rw`.  lol.

jasonmolenda commented:
@DavidSpickett pushed the fix for the `watch set expression` bug in 
```
commit 35e3939cb06d19942a30fd39f1416f49a7b982a4
Author: Jason Molenda <jmolenda@apple.com>
Date:   Thu Sep 21 14:48:19 2023 -0700

    watch set expression's default type was wrong with new modify type
```

jasonmolenda commented:
re. TestStepOverWatchpoint.py this test case annoys me greatly.  It wants to test that we get a watchpoint trap when stepping over a function that accesses the memory, and that we can stop on the watchpoint access when instruction stepping.  It wants to test this for read and write watchpoints.

It's operating on 1-byte globals, which are likely in the same dword as mentioned above.  MIPS and s390x both skip it expressedly because we're trying to watch the same word/dword of memory with both read & write access -- which isn't what the test is meant to be testing.

It starts by putting a read watchpoint on the read variable, and then for the case where it's doing write watchpoint testing, it adds a second watchpoint on the watch variable.  The read watchpoint is completely irrelevant in the watch-testing case.  None of this "read and write watch the same word of memory" is necessary!  The variables can be larger so they don't share memory granule, and we don't need to set a read watchpoint in the "test write-watchpoints" test method at all.

The test won't work on Darwin systems because hardware watchpoints/breakpoints are currently disabled while instruction stepping by the kernel.  So the bit of the test that instruction steps until a watchpoint is accessed will not succeed.

I'm going to leave this as expected fail for everyone for a tiny bit longer until I can test a bit of a rewrite.
